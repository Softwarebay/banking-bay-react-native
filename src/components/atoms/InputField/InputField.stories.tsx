import { boolean, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';

import { colors } from 'theme';
import { ViewStory } from 'utils/storybook/ViewStory';

import { InputField } from './InputField';

storiesOf('Atoms/InputField', module)
	.addDecorator((getStory) => (
		<ViewStory style={{ backgroundColor: colors.screenBackground }}>
			{getStory()}
		</ViewStory>
	))
	.add('Text', () => (
		<InputField
			placeholder={text('placeholder', 'Enter your mail')}
			secureTextEntry={boolean('secureTextEntry', false)}
			testID={text('testID', 'InputField-test-id')}
		/>
	))
	.add('Password', () => (
		<InputField
			placeholder={text('placeholder', 'Enter your password')}
			secureTextEntry={boolean('secureTextEntry', true)}
			testID={text('testID', 'InputField-test-id')}
		/>
	));
