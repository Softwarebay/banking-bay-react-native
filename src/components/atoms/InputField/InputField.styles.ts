import { StyleSheet } from 'react-native';

import { colors } from 'theme';

export const styles = StyleSheet.create({
	inputField: {
		color: colors.mainDark,
		fontFamily: 'MulishRegular',
		fontSize: 15,
	},
});
