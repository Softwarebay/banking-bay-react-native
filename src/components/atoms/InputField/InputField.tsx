import { forwardRef, useImperativeHandle, useState } from 'react';
import {
	KeyboardTypeOptions,
	StyleProp,
	TextInput,
	TextStyle,
} from 'react-native';

import { styles } from './InputField.styles';

interface Props {
	testID: string;
	placeholder?: string;
	secureTextEntry?: boolean;
	keyboardType?: KeyboardTypeOptions;
	style?: StyleProp<TextStyle>;
	validate?: (text: string) => boolean;
	setIsValid?: (valid: boolean) => void;
}

export interface InputFieldHandle {
	getValue: () => string;
	clear: () => void;
}

export const InputField = forwardRef<InputFieldHandle, Props>(
	(
		{
			testID,
			placeholder,
			secureTextEntry,
			keyboardType = 'default',
			style,
			validate,
			setIsValid,
		},
		ref
	) => {
		const [value, setValue] = useState('');

		useImperativeHandle(ref, () => ({
			getValue: () => value,
			clear: () => {
				setValue('');
			},
		}));

		const onChangeText = (text: string) => {
			setValue(text);
			if (validate && setIsValid) {
				const isValid = validate(text);
				setIsValid(isValid);
			}
		};

		return (
			<TextInput
				autoCorrect={false}
				keyboardType={keyboardType}
				placeholder={placeholder}
				secureTextEntry={secureTextEntry}
				style={[styles.inputField, style]}
				testID={testID}
				value={value}
				onChangeText={onChangeText}
			/>
		);
	}
);
