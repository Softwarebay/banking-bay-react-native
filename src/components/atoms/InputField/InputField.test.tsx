import { fireEvent, render, screen } from '@testing-library/react-native';
import testRenderer from 'react-test-renderer';

import { InputField } from './InputField';

import type { Props } from './InputField';

describe('InputField', () => {
	const props: Props = {
		testID: 'input-field-test-id',
		placeholder: 'Placeholder',
		style: {
			backgroundColor: 'red',
		},
	};

	it('should match snapshot', () => {
		const tree = testRenderer.create(<InputField {...props} />).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('should hide password', () => {
		render(<InputField {...props} secureTextEntry />);

		const input = screen.getByTestId('input-field-test-id');

		expect(input.props).toHaveProperty('secureTextEntry', true);
	});

	it('should display placeholder', () => {
		render(<InputField {...props} />);

		const input = screen.getByTestId('input-field-test-id');

		expect(input.props).toHaveProperty('placeholder', 'Placeholder');
	});

	it('should apply additional styles', () => {
		render(<InputField {...props} />);

		const input = screen.getByTestId('input-field-test-id');

		expect(input).toHaveStyle({
			backgroundColor: 'red',
		});
	});

	it('should handle type events', () => {
		render(<InputField {...props} />);

		const input = screen.getByTestId('input-field-test-id');

		fireEvent.changeText(input, 'Input text');

		expect(input.props).toHaveProperty('value', 'Input text');
	});
});
