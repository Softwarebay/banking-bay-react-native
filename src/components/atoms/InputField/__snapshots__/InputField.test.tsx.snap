// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField should match snapshot 1`] = `
<TextInput
  autoCorrect={false}
  keyboardType="default"
  onChangeText={[Function]}
  placeholder="Placeholder"
  style={
    Array [
      Object {
        "color": "#1B1D4D",
        "fontFamily": "MulishRegular",
        "fontSize": 15,
      },
      Object {
        "backgroundColor": "red",
      },
    ]
  }
  testID="input-field-test-id"
  value=""
/>
`;
